cip/CIP2015-05-13-EXISTS.adoc

```diff
@@ -0,0 +1,154 @@
+They are expressions, and as such must be side-effect free. That is, the subqueries in +EXIST+ must not be updating queries.
+
+==== Property exists
+
+In the property checking form, either an identifier or a string expression is used to represent the property key name. If the node/relationship returned has a property with the given name, the +EXISTS(...)+ expression should return true, otherwise it should return false. If the node/relationship expression returns a null value, the +EXISTS(...)+ expression should return null as well.
+
+This form of +EXISTS+ replaces the already existing +has(...)+ function.
+
+==== Subquery exists
+Identifiers and parameters from the outer query are available in the subquery. Identifiers introduced in the subquery are not available on the outside.
+
+For each matching subgraph evaluated with +EXISTS(...)+, the result value must be true if the subquery finds at least one matching row. If no matches are found, false should be returned.
+
+Regarding nomenclature, +EXISTS(...)+ is a scalar expression - it returns a single boolean value.
+
+==== Simple subquery exists
+When the subquery can be described with a single +MATCH-WHERE+ clause, the +MATCH+ keyword can be omitted.
+
+=== Examples
+_Example 1A:_
+
+Return all nodes that have a property named slogan.
+----
+MATCH (actor)
+WHERE EXISTS(actor.slogan)
+RETURN actor
+----
+
+_Example 1B:_
+----
+MATCH (actor)
+WHERE EXISTS(actor["slogan"])
+RETURN actor
+----
+
+
+_Example 2A:_
+
+Find all actors that have acted together with another actor with the same name.
+
+----
+MATCH (actor:Actor)
+WHERE EXISTS {
+	(other:Actor)-[:ACTED_IN]->(movie)<-[:ACTED_IN]-(actor)
+	WHERE other.name = actor.name
+}
+RETURN actor
+----
+
+_Example 2B:_
+
+Find all actors that have acted together with another actor with the same name.  +
+
+----
+MATCH (actor:Actor)
+WHERE EXISTS {
+	MATCH (other:Actor)-[:ACTED_IN]->(movie)<-[:ACTED_IN]-(actor)
+	WHERE other.name = actor.name
+}
+RETURN actor
+----
+
+=== Interaction with existing features
+This replaces the already existing +has(...)+ function.
+
+== What others do
+
+This is very close to what SQL does with its EXISTS functionality.
+
+This is also very similar in syntax to what SPARQL does with its EXISTS functionality; the rules regarding identifiers are identical, and the inner query also takes a subquery as input.
+
+== Benefits to this proposal
+
+The existing pattern predicate functionality is very useful, but does not cover all cases. Pattern predicates do not allow for introducing identifiers, which makes some queries - such as the one below - difficult to express succinctly:
+----
+
+MATCH (person:Person)
+WHERE EXISTS {
+  (person)-[:HAS_DOG]->(dog:Dog)
+  WHERE person.name = dog.name
+}
+RETURN person
+
+----
+This proposal also allows for powerful subqueries, for example using aggregation inside the EXISTS query. E.g. find all teams that have at least two members that have worked on projects considered successes.
+
```

Comment:
"least two members that have worked on projects considered successes." -> "least two members who have worked on successful projects."